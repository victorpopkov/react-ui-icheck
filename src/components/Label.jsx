import React, { Component } from 'react';
import PropTypes from 'prop-types';

class Label extends Component {
  constructor(props) {
    super(props);

    this.handleHover = this.handleHover.bind(this);
  }

  handleHover(event, isHovered = true) {
    const { disabled, handleHover, isMobile, labelHover } = this.props;

    if (disabled) return;

    if (handleHover && labelHover && !isHovered) {
      handleHover(event, false);
    } else if (handleHover) {
      handleHover(event, isHovered);
    }

    if (isMobile) event.stopPropagation();
  }

  renderLabel() {
    const { label, labelChildren, labelTag, labelTagClassName } = this.props;

    if (!React.isValidElement(label)) {
      const Tag = labelTag;

      if (labelChildren) {
        return <Tag className={labelTagClassName}>{labelChildren}</Tag>;
      }

      return (
        // eslint-disable-next-line react/no-danger
        <Tag
          className={labelTagClassName}
          dangerouslySetInnerHTML={{ __html: label }}
        />
      );
    }

    return null;
  }

  render() {
    const { children, labelClassName } = this.props;

    return (
      <label
        className={labelClassName || null}
        onBlur={(event) => this.handleHover(event, false)}
        onFocus={this.handleHover}
        onMouseOut={(event) => this.handleHover(event, false)}
        onMouseOver={this.handleHover}
        onTouchEnd={(event) => this.handleHover(event, false)}
        onTouchStart={this.handleHover}
      >
        {children}
        {this.renderLabel()}
      </label>
    );
  }
}

Label.propTypes = {
  children: PropTypes.node.isRequired,
  handleHover: PropTypes.func.isRequired,
  disabled: PropTypes.bool,
  isMobile: PropTypes.bool,
  label: PropTypes.node,
  labelChildren: PropTypes.node,
  labelClassName: PropTypes.string,
  labelHover: PropTypes.bool,
  labelTag: PropTypes.string,
  labelTagClassName: PropTypes.string,
};

Label.defaultProps = {
  disabled: false,
  isMobile: false,
  label: null,
  labelChildren: null,
  labelClassName: null,
  labelHover: true,
  labelTag: 'span',
  labelTagClassName: null,
};

export default Label;
