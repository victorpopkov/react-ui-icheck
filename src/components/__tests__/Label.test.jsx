import { mount, shallow } from 'enzyme';
import React from 'react';
import expect from 'expect';
import Label from '../Label';

describe('<Label /> component', () => {
  const testLabel = (
    <Label handleHover={() => false}>
      <div>child</div>
    </Label>
  );

  it('should have the default props values when mounted', () => {
    const wrapper = mount(testLabel);
    expect(wrapper.props()).toMatchSnapshot();
  });

  describe('should render the <label />', () => {
    it('when only the default props are passed', () => {
      const wrapper = shallow(testLabel);
      expect(wrapper).toMatchSnapshot();
    });

    it('when label as element is passed', () => {
      const wrapper = shallow(
        <Label handleHover={() => false} label={<span>propLabel</span>}>
          <div>child</div>
        </Label>,
      );

      expect(wrapper).toMatchSnapshot();
    });

    it('when label as string with labelTag and labelTagClassName are passed', () => {
      const wrapper = shallow(
        <Label
          handleHover={() => false}
          label="propLabel"
          labelTag="div"
          labelTagClassName="test"
        >
          <div>child</div>
        </Label>,
      );

      expect(wrapper).toMatchSnapshot();
    });

    it('when labelChildren, labelTag and labelTagClassName are passed', () => {
      const wrapper = shallow(
        <Label
          handleHover={() => false}
          labelChildren={<span>propLabelChildren</span>}
          labelTag="div"
          labelTagClassName="test"
        >
          <div>child</div>
        </Label>,
      );

      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('should handle the <label />', () => {
    const testSpy = jest.fn();
    const testSpyLabel = mount(
      <Label handleHover={testSpy}>
        <div>child</div>
      </Label>,
    );

    afterEach(() => {
      testSpy.mockClear();
    });

    describe('onFocus()', () => {
      it('with the default behaviour', () => {
        expect(testSpy).not.toHaveBeenCalled();
        testSpyLabel.simulate('focus');
        expect(testSpy).toHaveBeenCalledWith(expect.anything(), true);
      });

      it('when disabled', () => {
        const wrapper = mount(
          <Label handleHover={testSpy}>
            <div>child</div>
          </Label>,
        );

        expect(testSpy).not.toHaveBeenCalled();

        wrapper.setProps({ disabled: true });
        wrapper.simulate('focus');

        expect(testSpy).not.toHaveBeenCalled();
      });

      it('when on mobiles', () => {
        const wrapper = mount(
          <Label handleHover={testSpy} isMobile>
            <div>child</div>
          </Label>,
        );

        wrapper.simulate('focus');

        expect(testSpy).toHaveBeenCalled();
      });
    });

    it('onBlur()', () => {
      expect(testSpy).not.toHaveBeenCalled();
      testSpyLabel.simulate('blur');
      expect(testSpy).toHaveBeenCalledWith(expect.anything(), false);
    });

    it('onMouseOver()', () => {
      expect(testSpy).not.toHaveBeenCalled();
      testSpyLabel.simulate('mouseOver');
      expect(testSpy).toHaveBeenCalledWith(expect.anything(), true);
    });

    it('onMouseOut()', () => {
      expect(testSpy).not.toHaveBeenCalled();
      testSpyLabel.simulate('mouseOut');
      expect(testSpy).toHaveBeenCalledWith(expect.anything(), false);
    });

    it('onTouchStart()', () => {
      expect(testSpy).not.toHaveBeenCalled();
      testSpyLabel.simulate('touchStart');
      expect(testSpy).toHaveBeenCalledWith(expect.anything(), true);
    });

    it('onTouchEnd()', () => {
      expect(testSpy).not.toHaveBeenCalled();
      testSpyLabel.simulate('touchEnd');
      expect(testSpy).toHaveBeenCalledWith(expect.anything(), false);
    });
  });
});
