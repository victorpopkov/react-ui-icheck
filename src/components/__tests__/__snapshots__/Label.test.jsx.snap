// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Label component when the required \`handleHover\` prop is passed and the children are passed and the \`label\` prop is \`<span>propLabel</span>\` as shallow should match the snapshot 1`] = `
<label
  className={null}
  onBlur={[Function]}
  onFocus={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
  onTouchEnd={[Function]}
  onTouchStart={[Function]}
>
  <div>
    child
  </div>
</label>
`;

exports[`Label component when the required \`handleHover\` prop is passed and the children are passed and the \`labelTag\` prop is \`div\` and the \`labelTagClassName\` prop is \`test\` and the \`label\` prop is \`propLabel\` should match the snapshot 1`] = `
<label
  className={null}
  onBlur={[Function]}
  onFocus={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
  onTouchEnd={[Function]}
  onTouchStart={[Function]}
>
  <div>
    child
  </div>
  <div
    className="test"
    dangerouslySetInnerHTML={
      Object {
        "__html": "propLabel",
      }
    }
  />
</label>
`;

exports[`Label component when the required \`handleHover\` prop is passed and the children are passed and the \`labelTag\` prop is \`div\` and the \`labelTagClassName\` prop is \`test\` and the \`labelChildren\` prop is \`<span>propLabelChildren</span>\` should match the snapshot 1`] = `
<label
  className={null}
  onBlur={[Function]}
  onFocus={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
  onTouchEnd={[Function]}
  onTouchStart={[Function]}
>
  <div>
    child
  </div>
  <div
    className="test"
  >
    <span>
      propLabelChildren
    </span>
  </div>
</label>
`;

exports[`Label component when the required \`handleHover\` prop is passed and the children are passed as mount should match the default props snapshot 1`] = `
Object {
  "children": <div>
    child
  </div>,
  "disabled": false,
  "handleHover": [MockFunction],
  "isMobile": false,
  "label": null,
  "labelChildren": null,
  "labelClassName": null,
  "labelHover": true,
  "labelTag": "span",
  "labelTagClassName": null,
}
`;

exports[`Label component when the required \`handleHover\` prop is passed and the children are passed as shallow should match the snapshot 1`] = `
<label
  className={null}
  onBlur={[Function]}
  onFocus={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
  onTouchEnd={[Function]}
  onTouchStart={[Function]}
>
  <div>
    child
  </div>
  <span
    className={null}
    dangerouslySetInnerHTML={
      Object {
        "__html": null,
      }
    }
  />
</label>
`;
